Client/trunk/ParaEngineClient/renderer/effect_file.h

```diff
@@ -303,6 +309,32 @@ namespace ParaEngine
 		bool SetTechniqueByIndex(int nIndex) { return false; };
 		/** get the current technique description. This function may return NULL*/
 		const TechniqueDesc* GetCurrentTechniqueDesc() { return NULL; };
+
+		void onDrawPass(CParameterBlock* pMaterialParams,int passIndex)
+		{
+			ScriptCallback* pCallback=GetScriptCallback(Type_DrawPass);
+			if(pCallback)
+			{
+				thread_local static string code;
+				code="msg={";
```

Comment:
if you want better performance, do following:
instead of msg={}, which create a temporary table each call, change to:
```
pass_index = 1;
XXXXXX();
```
GC for temporary object is a skiller to framerate. I have tried hard to ensure 0 memory allocate (new tables) during each frame.

Reply:
for callbackkey, better encode in XXXXX(some_callback_key);